### PR TITLE
Don't call on_dieplayer callbacks two times

### PR DESCRIPTION
**Pull request by savilli**
"When you fix one bug you make two" - savilli 2021 

After #11680 on_dieplayer callbacks may be called two times, which was fixed in #11411.
A player dies, on_dieplayer callbacks are called, but the player's client isn't aware of it yet and sends fail damage.
on_dieplayer callbacks are called second time here https://github.com/minetest/minetest/blob/0c4929f0252811d9715b96379973929053d32ef0/src/network/serverpackethandler.cpp#L830

We need better checks.

## To do
This PR is Ready for Review.

## How to test
It's hard to reproduce this naturally, just follow the code.

### CODE REVIEW

File: src/server/player_sao.cpp [hunk 1]
```diff
@@ -473,26 +473,31 @@ void PlayerSAO::setHP(s32 hp, const PlayerHPChangeReason &reason, bool send)
...
 	m_hp = hp;
 
 	// Update properties on death
 	if ((hp == 0) != (oldhp == 0))
 		m_properties_sent = false;
 
-	if (send)
-		m_env->getGameDef()->SendPlayerHPOrDie(this, reason);
+	m_env->getGameDef()->SendPlayerHPOrDie(this, reason);
 }
```
Comment (sfan5) [inline src/server/player_sao.cpp:501]: I think it'd work out better if you rewrote this function so that it does not return early.

Then at the end you can have central handling that looks like this:
```c
if (from_client) {
    if (client_expects_hp != actual_new_hp) // Correct client side prediction
        SendPlayerHPOrDie();
} else {
    if (old_hp != actual_new_hp)
        SendPlayerHPOrDie();
}
```
This would be much clearer and can be confirmed bug-free from a single look.

Comment (savilli) [reply src/server/player_sao.cpp:501]: > `if (client_expects_hp != actual_new_hp)`

If client_expects_hp == 0 and actual_new_hp == 0, you still should send death screen.

File: src/server/player_sao.cpp [hunk 2]
```diff
@@ -463,41 +463,32 @@ void PlayerSAO::rightClick(ServerActiveObject *clicker)
...
+	m_hp = old_hp;
 
 	// Update properties on death
-	if ((hp == 0) != (oldhp == 0))
+	if ((hp == 0) != (old_hp == 0))
 		m_properties_sent = false;
 
-	m_env->getGameDef()->SendPlayerHPOrDie(this, reason);
+	if (from_client ? (hp != orig_hp || hp == 0) : (hp != old_hp))
+		m_env->getGameDef()->SendPlayerHPOrDie(this, reason);
```
Comment (sfan5) [inline src/server/player_sao.cpp]: Readability still matters
```suggestion
	if (from_client) {
		// Send hp if it doesn't match the client side prediction OR if player died
		if (hp != orig_hp || hp == 0)
			m_env->getGameDef()->SendPlayerHPOrDie(this, reason);
	} else {
		if (hp != old_hp)
			m_env->getGameDef()->SendPlayerHPOrDie(this, reason);
	}
```

File: src/server.cpp [hunk 3]
```diff
@@ -1359,18 +1360,20 @@ void Server::SendMovement(session_t peer_id)
...
-		SendPlayerHP(peer_id);
-	else
-		DiePlayer(peer_id, reason);
+	if (playersao->isDead())
+		HandlePlayerDeath(playersao, reason);
+}
+
+void Server::SendPlayerHP(PlayerSAO *playersao) {
+	SendHP(playersao->getPeerID(), playersao->getHP());
 }
```
Comment (sfan5) [inline src/server.cpp:1377]:
```suggestion
void Server::SendPlayerHP(PlayerSAO *playersao)
{
	SendHP(playersao->getPeerID(), playersao->getHP());
}
```

**Review by sfan5 — COMMENTED** [review]
Spent 10 mins staring at the conditions and looks correct, will test later

File: src/server/player_sao.cpp [hunk 4]
```diff
@@ -463,36 +463,34 @@ void PlayerSAO::rightClick(ServerActiveObject *clicker)
...
-		return; // Cannot take more damage
+	if (hp == m_hp)
+		return; // Nothing to do
 
-	{
-		s32 hp_change = m_env->getScriptIface()->on_player_hpchange(this, hp - m_hp, reason);
-		if (hp_change == 0)
-			return;
+	s32 hp_change = m_env->getScriptIface()->on_player_hpchange(this, hp - (s32)m_hp, reason);
+	hp_change = rangelim(hp_change, -U16_MAX, U16_MAX);
```
Comment (sfan5) [inline src/server/player_sao.cpp:474]: if it's rangelimited anyway doing this here is pointless

Comment (savilli) [reply src/server/player_sao.cpp:474]: Do you mean `hp = rangelim(hp, 0, U16_MAX);` ? Well, on_player_hpchange can return 2147483647 and it will overflow.

Comment (sfan5) [reply src/server/player_sao.cpp:474]: The `hp_change = rangelim(hp_change, -U16_MAX, U16_MAX);` is pointless.
Sure it can protect against integer overflow but I don't think that's really important here.

Comment (savilli) [reply src/server/player_sao.cpp:474]: It won't hurt to protect against stupidness

Comment (savilli) [reply src/server/player_sao.cpp:474]: We have the same check there https://github.com/minetest/minetest/blob/1bc753f655db4c7de030f2700361011e1c0278a5/src/script/lua_api/l_item.cpp#L105-L106 for example

Comment (sfan5) [reply src/server/player_sao.cpp:474]: That's different. Anyway it doesn't matter.

File: src/script/cpp_api/s_player.cpp [hunk 5]
```diff
@@ -109,9 +109,9 @@ s32 ScriptApiPlayer::on_player_hpchange(ServerActiveObject *player,
 
 	// Call callbacks
 	PCALL_RES(lua_pcall(L, 3, 1, error_handler));
-	hp_change = lua_tointeger(L, -1);
+	lua_Integer change = lua_tointeger(L, -1);
```
Comment (sfan5) [inline src/script/cpp_api/s_player.cpp:112]: `s32` is already an int, so is `lua_Integer`
just don't

**Comment by sfan5** [comment]
Tested with
```lua
minetest.register_on_player_hpchange(function(player, hp_change, reason)
	if reason.type == "fall" then
		return 0
	end
	return hp_change
end, true)
```
still died. Wouldn't recommend.

File: src/script/cpp_api/s_player.cpp (hunk 5 above)
Comment (savilli) [reply src/script/cpp_api/s_player.cpp:112]: Alright, let it be s32

**Comment by savilli** [comment]
> Tested with
> 
> ```lua
> minetest.register_on_player_hpchange(function(player, hp_change, reason)
> 	if reason.type == "fall" then
> 		return 0
> 	end
> 	return hp_change
> end, true)
> ```
> 
> still died. Wouldn't recommend.

Fixed

**Review by sfan5 — APPROVED** [review]
Works.

File: src/server/player_sao.cpp [hunk 6]
```diff
@@ -463,36 +463,33 @@ void PlayerSAO::rightClick(ServerActiveObject *clicker)
 	m_env->getScriptIface()->on_rightclickplayer(this, clicker);
 }
 
-void PlayerSAO::setHP(s32 hp, const PlayerHPChangeReason &reason, bool send)
+void PlayerSAO::setHP(s32 orig_hp, const PlayerHPChangeReason &reason, bool from_client)
```
Comment (SmallJoker) [inline src/server/player_sao.cpp:466]: "original" may mislead to assume that it's the PlayerSAO's original HP.
Why is this rename necessary? This variable becomes unused after `hp`'s initialization. If you'd like to keep the variables separated, I would suggest `new_hp` or `target_hp`.

**Review by SmallJoker — APPROVED** [review]
Works **but**

The comment above should be handled before merge.

File: src/server/player_sao.cpp (hunk 6 above)
Comment (savilli) [reply src/server/player_sao.cpp:466]: `target_hp` is good
